### ISSUE TITLE
Zone Layout and FancyZones integration

### ISSUE THREAD

**Issue by RamonUnch** [RamonUnch/AltSnap#97]
@child-duckling and @shelbyKiraM
I have been working for some time on some snapping layouts like you have on FancyZones.
Here is a beta version of AltDrag and hooks.dll that supports this feature.

at the end of your ini file add:
```
[Zones]
UseZones=1
; Set to 1 to enable Zone Layouts

FancyZone=1
; Set to 1 for better FancyZone integration (x64 builds only!).

Zone0=425,0,953,627,
Zone1=0,627,953,1050,
Zone2=0,0,425,627,
Zone3=953,0,1920,1050,
; List of zones coordinates LEFT,TOP,RIGHT,BOTTOM,
; Up to 32 zones and all zones ends with a coma
```

To setup the zones you can (for now), open several test windows, (Advanced tab of the config)
Then, once you placed them like you please on the desktop, you can go into the Tray context-menu and click on the "Save Zones" option.
So I suggest you snap them to your FancyZones grid then save them to AltSnap.

Like on PowerToys you have to use Shift in order to Snap to a Zone.
if you point close to the limit between two points , then the window is snapped to the "union" rectangle.

EDIT Latest test build:
32bit builds for i386 and later: [AltSnap1.47test25bin.zip](https://github.com/RamonUnch/AltSnap/files/7250360/AltSnap1.47test25bin.zip)
64 bit builds for FancyZone integration:[AltSnap1.47test25bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7250361/AltSnap1.47test25bin_x64.zip)
Installers: [AltSnap1.47test25-installers.zip](https://github.com/RamonUnch/AltSnap/files/7250362/AltSnap1.47test25-installers.zip)

For FancyZones integration 64bit AltSnap is necessary!

see https://github.com/RamonUnch/AltSnap/issues/13, https://github.com/RamonUnch/AltSnap/issues/80 and https://github.com/RamonUnch/AltSnap/issues/89 for more discussion

**Comment by RamonUnch** [comment]
[AltSnap1.47-test11bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7168855/AltSnap1.47-test11bin_x64.zip)
Here is an improved version.
* After enabling the `UseZones` option, you will be able to open new Test Windows directly from the tray menu.
* When saving the Zone layout, all test windows will be closed.
* The previous Zone layout will be cleared before saving the new one.
* Only One layout is supported. This will remain this way for a while.

The point is that if you setup a snap layout you like once and then rarely change it.
If you use PowerToys, you *should* set `FancyZone=1`, then create the Layout you like with PowerToys, then import this layout to AltDrag by snapping the correct amount of **Test Windows** to the FancyZones (using Powertoys). Then hit the **Save test windows as zones** entry in the tray menu.

If you do not want to *AltSnap* to zones, then you do not have to do this and you can set `UseZones=0`, and still set `FancyZone=1` if you Want AltSnap to restore FancyZones-Snapped windows and vice-versa.
For this interaction it is necessary to use the x64 build of AltDrag as the windows dimensions are stored by FancyZones in a 64bit pointer that directly contains the dimensions in two halfs of 32b each `(width|height<<32)`, This pointer is stored in a property of the window and such pointers are truncated when accessing them from a 32b process.

**Comment by RamonUnch** [comment]
[AltSnap1.47-test12bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7176324/AltSnap1.47-test12bin_x64.zip)
* Fix restore flags not set properly when using `FancyZone=1`
* Added Ctrl to be able to extend windows over several zones, like with FancyZones

**Comment by RamonUnch** [comment]
More fixes and minor adjustements:
[AltSnap1.47-test13bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7179448/AltSnap1.47-test13bin_x64.zip)

And also 32 bits builds for good measure (no FancyZone integration) but with snap layouts support
[AltSnap1.47-test13bin_i386.zip](https://github.com/RamonUnch/AltSnap/files/7179467/AltSnap1.47-test13bin_i386.zip)

**Comment by Ichisich** [comment]
Currently, the InterZone option leads to temporary maximized window between zones.
During slow dragging it's clearly visible and fast dragging introduces a short, but noticeable, flicker.

Nevertheless, a surprisingly robust feature (UseZones), especially for such an early draft.

**Comment by RamonUnch** [comment]
The InterZone option was previously hard coded to 32, so I am not sure it is related to this.
I would not have expected maximization, though.

Was it when you were using negative InterZone values? With negative values, if the pointer is in a margin between two zones, then the window is restored to its last non-snapped size (not maximized though), small negative values for InterZone will lead to flicker between zones.

Also if you point the cursor close to the junction between 2, 3 or 4 zones, it will be resized, so that the window covers all zones completely (union rectangle) depending on your layout it may correspond to the whole screen. FancyZones should behave similarly.

Also for now when using CTRL, and move pixel perfect the cursor vertically or horizontally, then the window will not be snapped properly (because i use the IntersectRect function that ignores RECTs with empty content. I will re-implement it better.

I hope to be close from release, so from now only minor adjustements should be done.

**Comment by Ichisich** [comment]
> Was it when you were using negative InterZone values?
No, the default 32.

> Also if you point the cursor close to the junction between 2, 3 or 4 zones, it will be resized, so that the window covers all zones completely (union rectangle) depending on your layout it may correspond to the whole screen.
**That should be the culprit.**
I tested some other zone constellations and in the interzone the window always takes the shape of the summarized zones I'm moving in-between (InterZone=32).
**With InterZone=0 the window snaps instantaneous without in-between shapes.**

> Also for now when using CTRL, and move pixel perfect the cursor vertically or horizontally, then the window will not be snapped properly
Couldn't replicate a misalignment (smartaero=7/ SnapThreshold=20).
Do you mean CTRL as general meta key for dragging or extending over multiple zones via meta+shitft+ctrl?

**Comment by RamonUnch** [comment]
Thanks for your feedback.

>Couldn't replicate a misalignment (smartaero=7/ SnapThreshold=20).
Do you mean CTRL as general meta key for dragging or extending over multiple zones via meta+shitft+ctrl?

It is very rare and happened on a very lowres screen to me.
When you hit ctrl the point is saved, then when you move your cursor AltSnap calculate the rect between the initial point and the current cursor point. If the width or the height of this rectangle is exactly zero then the window is restored because the IntersectRect finds a zero-area intersection rectangle and thus fails. It is indeed difficult to reproduce, I was able only a few times.

In your case InterZone=0 is a good value, the option is here for people that want to extend the window to several zones without hitting Ctrl, but it is less powerful than Ctrl of course.

**Comment by RamonUnch** [comment]
> Do you mean CTRL as general meta key for dragging or extending over multiple zones via meta+shitft+ctrl?

Yes exactly, like with FancyZones,
I was not clear, once you hit ctrl and you move your cursor, all zones that are touching the rectangle defined by the  initial ctrl point to the cursor point become covered.
This is useful for exemple if you use many small zones and often want to extend a window over several zones.
I initially thought I would focus on snapping to a single zone, but then I saw that extending was not very difficult.

**Comment by Ichisich** [comment]
Understood, and I'm finally able to reproduce it reliably.

Grab a window near its edge, to shorten the distance to the next zone, and drag in a perfect horizontal/vertical manner to draw an imaginary line instead of a rectangle and the window doesn't snap into the new zone but gets just dragged.

I doubt this will be a serious problem. Nobody drags a window this slowly to draw a perfect line with an ordinary mouse and a 1080p+ monitor. Most people flick their mouse, thus produce a pseudo rectangle by accident.

**Comment by RamonUnch** [comment]
Actually I simplified the code and do like FancyZones does by just using the union between the first zone and the last zone of the Ctrl+Alt+Shift+click-drag selection. This is also more desirable in my opinion as it extends less excessively the window when you expend it over many zones.
64 bit builds for FancyZone integration: [AltSnap1.47-test14bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7183593/AltSnap1.47-test14bin_x64.zip)
32bit builds for i386 and later: [AltSnap1.47-test14bin_i386.zip](https://github.com/RamonUnch/AltSnap/files/7183592/AltSnap1.47-test14bin_i386.zip)

**Comment by frozensnowy** [comment]
Hey, thanks for this! 

Unfortunately though, it doesn't seem to work properly on my monitor - which has a resolution of 5120x1440. The snap layouts are all weirdly squished and stuck in the top left corner. 

It looks as if it's stuck to a 1920x1080 resolution for some odd reason. Moving outside of that, it returns to the standard Windows aero snap layout. It seems to work well otherwise though.

In case it helps, I'm using the version you posted just now.

**Comment by RamonUnch** [comment]
The default snap layout is **bad** this is normal, it is here just for test purposes.
You have to create your own by making several test windows with the Tray menu: Position the test windows as you like and then again from the tray menu, save hit the Save test windows as new snap layout then hit OK.

**Comment by RamonUnch** [comment]
Do you mean that you often change your screen resolution?
The idea is that snap layouts are user defined and no default snap layouts are supposed to exist.

**Comment by frozensnowy** [comment]
> 
> 
> The default snap layout is **bad** this is normal, it is here just for test purposes.
> You have to create your own by making several test windows with the Tray menu: Position the test windows as you like and then again from the tray menu, save hit the **Save test windows as snap layout** then hit OK.

Ahh yeah that worked well, thanks! I appreciate the help



> 
> 
> Do you mean that you often change your screen resolution?
> The idea is that snap layouts are user defined and no default snap layouts are supposed to exist.

I don't change resolution often, so it's all good

**Comment by shelbyKiraM** [comment]
Uh, I get a crash on Keyboard > Hotkeys when I disable all the options, leaving Modifier Key dropdown set to "Nothing" for the action. Luckily, my keyboard doesn't have a Right Windows Key,  but odd bug. I'm on Win11. The dropdown doesn't appear to affect it. But unchecking all the boxes here and applying crashes:
![image](https://user-images.githubusercontent.com/202692/133854143-d6b2f561-44f9-47a7-9e18-72e88a5bb621.png)
Is there a log file I can keep my eye on or a flag to start the program with, to output more debugging info?

**Comment by shelbyKiraM** [comment]
Ahhh, it's actually crashing when I change any setting. I think I found the bug, I must've edited the `.ini` config file with Atom and it saved it in a non-binary (hahahah that's my gender) (prob ASCII or UTF-8?) format or something? Gonna restore the original over it and only edit THAT in `notepad.exe`.

**Comment by RamonUnch** [comment]
.ini file is supposed to be encoded in UTF-16 not utf8 but, I did encounter random crash sometime when changing config, so maybe this is not the only problem. I should check on the code with a debugger.
I rarely had those crash though and it was never reproducible though... I hop file encoding is the main problem.

**Comment by shelbyKiraM** [comment]
Looks like it was bugging out when I disabled all hotkeys... there might be something else I'm missing tho, cuz I had it in a state where it would crash whenever I changed many settings in Mouse and Keyboard.

**Comment by shelbyKiraM** [comment]
This is the bad ini:
[AltSnap.ini.txt](https://github.com/RamonUnch/AltSnap/files/7188920/AltSnap.ini.txt)
 renamed to let me upload here. Diffing it with the default one may give some clue to the issue, too. The layout looks all weird in Notepad, dunno.

**Comment by shelbyKiraM** [comment]
I DID erase the old `%APPDATA%\AltDrag` dir I used before and copied the zip output to get it back working, though.

**Comment by RamonUnch** [comment]
Yes in the above linked txt file, the encoding is UTF-16LE as it should, however the stream prefix (seen with a hex editor) is `EF BB BF` indicating UTF-8 encoding.
This is a bug from the text editor you used, the correct prefix for UTF-16 streams is `FF FE` Which txt editor are you using? it seems it thinks clever to always add UTF-8 prefix even when encoding is different! maybe you enabled some options or something.

If the prefix is wrong, then windows can no longer determine the proper encoding and tries to read utf-16 like if it was utf-8, so some function seems to be crashing, also it seems Notepad converts null characters into spaces and removes line breaks, even Notepad++ is unable to read the file properly and displays `NULL` every other characters.

**Comment by shelbyKiraM** [comment]
https://Atom.io - GitHub's own editor hehe. Maybe it's time to make an issue in there.

**Comment by RamonUnch** [comment]
> https://Atom.io - GitHub's own editor hehe. Maybe it's time to make an issue in there.

Did you try to reproduce the bug? It looks very strange to me from such a good text editor for which all Unicode encoding support is for sure very important.

**Comment by RamonUnch** [comment]
Minor updates:
i386 builds [AltSnap1.47test15bin.zip](https://github.com/RamonUnch/AltSnap/files/7196921/AltSnap1.47test15bin.zip)
x86_64 builds [AltSnap1.47test15bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7196922/AltSnap1.47test15bin_x64.zip)

**Comment by RamonUnch** [comment]
EDIT Important Crash fix:
64 bit builds for FancyZone integration: [AltSnap1.47test16bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7202082/AltSnap1.47test16bin_x64.zip)
32bit builds for i386 and later: [AltSnap1.47test16bin.zip](https://github.com/RamonUnch/AltSnap/files/7202081/AltSnap1.47test16bin.zip)

**Comment by RamonUnch** [comment]
@shelbyKiraM
Actually I tested the file and the Atom bug is just a configuration quirk.
When you open an UTF-16 encoded file, by default Atom still displays it like if it was UTF-8, If you save then it saves the bad version. 
To open it correctly, before modifying, you have to select the proper encoding in the bottom bar. you can click on the UTF-8 bottom button, then  select Auto Detect from the popup menu.

So, I guess they will ignore your bug report, unless you re-formulate it:
It would be nice to have a way so that Atom always auto-detects the file encoding when opening it instead of always displaying as UTF-8.

**Comment by RamonUnch** [comment]
Test builds with GUI options.
64 bit builds for FancyZone integration: [AltSnap1.47test19bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7223889/AltSnap1.47test19bin_x64.zip)
32bit builds for i386 and later: [AltSnap1.47test19src.zip](https://github.com/RamonUnch/AltSnap/files/7223892/AltSnap1.47test19src.zip)

**Comment by shelbyKiraM** [comment]
> @xzayn
> Does Fancyzone integration also mean windows11 snapping support?

Not really, but you can set up zones in the four quarters to get very similar. It DOES still support alt-dragging to corners of the screen to snap like Aero Snap. There's some advanced layouts that Win11 Snap allows, it doesn't INTEGRATE with those, but they work well together, as someone who is has been on Win11 on both desktop & laptop for a fair time.

**Comment by RamonUnch** [comment]
@xzayn
I have seen how Win11 snap layouts are supposed to work, it is as shelbyKiraM said, not perfectly integrated with Win11 snapping.
What the Fancyzone integration does is simply makes AltSnap restore windows that were snapped to zones using FancyZones toom from MS Powertoys, so it has no uses outside of interfacing with FancyZones (do not bother if you do not use this latter)

What you could do to have better integration between all kind of snapping is also set **Restore AltSnapped windows with normal movet**. Unless you use the **Use AltSnap for normal title bar movement** and in this case, it does not matter as all kind of snapping should be handled by AltSnap.

Normally when a window is snapped by win11, then moving it with AltSnap will restore it.

**Comment by shelbyKiraM** [comment]
Actually, if anything, it improves AltDrag's snapping capabilities, almost entirely obviating my need for PowerToys. But I use more than FancyZones sometimes, so I still have it. 😁

**Comment by shelbyKiraM** [comment]
In config, it says "Use Snap Layout (configure with tray menu)" but there's nothing about setting up Snap in here: 
![image](https://user-images.githubusercontent.com/202692/134746013-d625968f-f57e-41f8-aa15-6a50d068916a.png)

**Comment by RamonUnch** [comment]
You are right, the Test Window has to be opened before the save layout option is displayed.
I will always display the save option grayed and ungray it if ant test window is opened.

**Comment by RamonUnch** [comment]
Updated build with more minor fixes, soon to be released:
Builds for i386: [AltSnap1.47test25bin.zip](https://github.com/RamonUnch/AltSnap/files/7250360/AltSnap1.47test25bin.zip)
Builds for x86_64b [AltSnap1.47test25bin_x64.zip](https://github.com/RamonUnch/AltSnap/files/7250361/AltSnap1.47test25bin_x64.zip)
Installers: [AltSnap1.47test25-installers.zip](https://github.com/RamonUnch/AltSnap/files/7250362/AltSnap1.47test25-installers.zip)

**Comment by RamonUnch** [comment]
Released 1.47, use this one for more issues
https://github.com/RamonUnch/AltSnap/releases
